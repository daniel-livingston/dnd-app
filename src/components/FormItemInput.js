import React from "react";
import PropTypes from "prop-types";

const propTypes = {
	type: PropTypes.string.isRequired,
	value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
	onChange: PropTypes.func.isRequired,
	placeholder: PropTypes.string.isRequired,
	options: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
};

const FormItemInput = (props) => {
	const getInput = (type, onChange, placeholder, value, options) => {
		switch (type) {
			case "number": {
				return (
					<input
						className='FormItemInput FormItemInput-Number'
						onChange={props.onChange}
						type='number'
						value={props.value}
					/>
				);
			}
			case "password": {
				return (
					<input
						className='FormItemInput FormItemInput-Password'
						onChange={onChange}
						placeholder={placeholder}
						type='text'
						value={value}
					/>
				);
			}
			case "select": {
				if (!options) throw Error("No options array provided");
				return (
					<select
						className='FormItemInput FormItemInput-Select'
						onChange={onChange}
						value={value}
					>
						{options.map((option) => (
							<option key={option} value={option}>
								{option}
							</option>
						))}
					</select>
				);
			}
			case "text": {
				return (
					<input
						className='FormItemInput FormItemInput-Text'
						onChange={onChange}
						placeholder={placeholder}
						type='text'
						value={value}
					/>
				);
			}
		}
	};

	return getInput(props.type, props.onChange, props.placeholder, props.value, props.options);
};

FormItemInput.propTypes = propTypes;

export default FormItemInput;
