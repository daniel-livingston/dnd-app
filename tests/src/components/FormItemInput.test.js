import React from "react";
import { shallow } from "enzyme";
import FormItemInput from "../../../src/components/FormItemInput";

const props = {
	type: "text",
	value: "Test value",
	onChange: jest.fn(),
	placeholder: "Test placeholder",
};

describe("<FormItemInput>", () => {
	test("renders number input correctly", () => {
		const wrapper = shallow(<FormItemInput {...props} type={"number"} value={1} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("renders password input correctly", () => {
		const wrapper = shallow(<FormItemInput {...props} type={"password"} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("renders select element correctly", () => {
		const wrapper = shallow(
			<FormItemInput
				{...props}
				type={"select"}
				value='Test option 1'
				options={["Test option 1", "Test option 2"]}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});

	test("renders text input correctly", () => {
		const wrapper = shallow(<FormItemInput {...props} />);
		expect(wrapper).toMatchSnapshot();
	});
});
