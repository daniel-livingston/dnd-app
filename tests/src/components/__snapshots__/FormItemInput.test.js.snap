// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormItemInput> renders number input correctly 1`] = `
<input
  className="FormItemInput FormItemInput-Number"
  onChange={[MockFunction]}
  type="number"
  value={1}
/>
`;

exports[`<FormItemInput> renders password input correctly 1`] = `
<input
  className="FormItemInput FormItemInput-Password"
  onChange={[MockFunction]}
  placeholder="Test placeholder"
  type="text"
  value="Test value"
/>
`;

exports[`<FormItemInput> renders select element correctly 1`] = `
<select
  className="FormItemInput FormItemInput-Select"
  onChange={[MockFunction]}
  value="Test option 1"
>
  <option
    key="Test option 1"
    value="Test option 1"
  >
    Test option 1
  </option>
  <option
    key="Test option 2"
    value="Test option 2"
  >
    Test option 2
  </option>
</select>
`;

exports[`<FormItemInput> renders text input correctly 1`] = `
<input
  className="FormItemInput FormItemInput-Text"
  onChange={[MockFunction]}
  placeholder="Test placeholder"
  type="text"
  value="Test value"
/>
`;
